### DISCUSSION TITLE
How to use rowSpan?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [TanStack/table#2233] [General] [answered]
I made a minimal repro of my original code. Here it is in HTML:

```html
<!DOCTYPE html>
<html>
<head>
<style>
table, th, td { 
  text-align: center;
  border: 1px dashed lightpink;
  color: navy;
}

</style>
</head>
<body>

<table>
  <tr>
    <th>Actor</th>
    <th>Movies</th>
  </tr>
  <tr>
    <td rowspan="4">Johnny Depp</td>
    <td>Pirates of the Carribean  1</td>
  </tr>
  <tr>
    <td>Pirates of the Carribean  2</td>
  </tr>
  <tr>
    <td>Pirates of the Carribean  3</td>
  </tr>
  <tr>
    <td>Pirates of the Carribean  4</td>
  </tr>
</table>

</body>
</html>
```

The output looks like:

![rowspan](https://user-images.githubusercontent.com/16436270/80309396-6dab4e00-87f2-11ea-9611-c9f4bbdad78a.PNG)

So I tried implementing it in [CodeSandBox](https://codesandbox.io). Following is the code:

```ts
import * as React from "react";
import { useTable } from "react-table";

const borderStyle = {
  border: "1px dashed navy"
};

export default function App() {
  const data = React.useMemo(
    () => [
      {
        actor: "Johnny Depp",
        movies: [
          {
            name: "Pirates of the Carribean 1"
          },
          {
            name: "Pirates of the Carribean 2"
          },
          {
            name: "Pirates of the Carribean 3"
          },
          {
            name: "Pirates of the Carribean 4"
          }
        ]
      }
    ],
    []
  );
  const columns = React.useMemo(
    () => [
      {
        Header: "Actor",
        accessor: "actor",
        enableRowSpan: true
      },
      {
        Header: "Movies",
        accessor: (row, index) => {
          console.log({ row });
          return row.movies.map(movie => movie.name);
        }
      }
    ],
    []
  );
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow
  } = useTable({ columns, data });
  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map(headerGroup => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <th {...column.getHeaderProps()} style={borderStyle}>
                {column.render("Header")}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map((row, i) => {
          prepareRow(row);
          if (i == 0) {
            console.log({ row });
          }
          return (
            <tr {...row.getRowProps()}>
              {row.cells.map((cell, j) => {
                if (i == 0 && j < 2) {
                  console.log({ cell, i, j });
                }
                return (
                  <td
                    rowSpan={cell.rowSpan}
                    {...cell.getCellProps()}
                    style={borderStyle}
                  >
                    {cell.render("Cell")}
                  </td>
                );
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}
```

Here's the direct link to it: https://codesandbox.io/s/modest-sanderson-z0keq?file=/src/App.tsx

It looks like:

![codesandbox](https://user-images.githubusercontent.com/16436270/80309485-00e48380-87f3-11ea-8040-9c08f4c2e866.PNG)

I want movie names to be one below the other but can't seem to figure it out 🤔

I want it to look like the HTML. The data shouldn't change as my original use-case fetches from an API. How do I do it?

**Comment by nleve** [comment]
There is no rowspan hook included in react-table currently. So while you are setting rowSpanEnabled on the header and referencing cell.rowSpan, you aren't ever setting cell.rowSpan or checking rowSpanEnabled again.

**Comment by deadcoder0904** [comment]
@liberza I got that. I left that part out purposely because Idk how to display the movie names row by row.

My movie list is an array of objects so how will I display it beside actor name? So it looks like:

| Actor       | Movies                     |
|-------------|----------------------------|
| Johnny Depp | Pirates of the Carribean 1 |
| Johnny Depp | Pirates of the Carribean 2 |
| Johnny Depp | Pirates of the Carribean 3 |
| Johnny Depp | Pirates of the Carribean 4 |

Once I fix that, I can totally see myself copying your [Codesandbox](https://codesandbox.io/s/dreamy-neumann-3ej9z?file=/src/App.js) ;)

**Comment by deadcoder0904** [comment, marked as answer]
Okay made it work. Guess I just had to flatten the data. It was as easy as just copying & pasting your code. Here's the working solution:

```ts
import * as React from "react";
import { useTable } from "react-table";

type Data = {
  actor: string;
  movie: string;
};

const borderStyle = {
  border: "1px solid gray",
  padding: "8px 10px"
};

function useInstance(instance) {
  const { allColumns } = instance;

  let rowSpanHeaders = [];

  allColumns.forEach((column, i) => {
    const { id, enableRowSpan } = column;

    if (enableRowSpan !== undefined) {
      rowSpanHeaders = [
        ...rowSpanHeaders,
        { id, topCellValue: null, topCellIndex: 0 }
      ];
    }
  });

  Object.assign(instance, { rowSpanHeaders });
}

export default function App() {
  const origData = [
    {
      actor: "Johnny Depp",
      movies: [
        {
          name: "Pirates of the Carribean 1"
        },
        {
          name: "Pirates of the Carribean 2"
        },
        {
          name: "Pirates of the Carribean 3"
        },
        {
          name: "Pirates of the Carribean 4"
        }
      ]
    }
  ];
  const newData: Array<Data> = [];
  origData.forEach(actorObj => {
    actorObj.movies.forEach(movie => {
      newData.push({
        actor: actorObj.actor,
        movie: movie.name
      });
    });
  });
  const data = React.useMemo(() => newData, []);
  const columns = React.useMemo(
    () => [
      {
        Header: "Actor",
        accessor: "actor",
        enableRowSpan: true
      },
      {
        Header: "Movies",
        accessor: "movie"
      }
    ],
    []
  );
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    rowSpanHeaders
  } = useTable({ columns, data }, hooks => {
    hooks.useInstance.push(useInstance);
  });
  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map(headerGroup => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <th {...column.getHeaderProps()} style={borderStyle}>
                {column.render("Header")}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map((row, i) => {
          prepareRow(row);

          for (let j = 0; j < row.allCells.length; j++) {
            let cell = row.allCells[j];
            let rowSpanHeader = rowSpanHeaders.find(
              x => x.id === cell.column.id
            );

            if (rowSpanHeader !== undefined) {
              if (
                rowSpanHeader.topCellValue === null ||
                rowSpanHeader.topCellValue !== cell.value
              ) {
                cell.isRowSpanned = false;
                rowSpanHeader.topCellValue = cell.value;
                rowSpanHeader.topCellIndex = i;
                cell.rowSpan = 1;
              } else {
                rows[rowSpanHeader.topCellIndex].allCells[j].rowSpan++;
                cell.isRowSpanned = true;
              }
            }
          }
          return null;
        })}
        {rows.map(row => {
          return (
            <tr {...row.getRowProps()}>
              {row.cells.map(cell => {
                if (cell.isRowSpanned) return null;
                else
                  return (
                    <td
                      style={borderStyle}
                      rowSpan={cell.rowSpan}
                      {...cell.getCellProps()}
                    >
                      {cell.render("Cell")}
                    </td>
                  );
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}
```

![react-table-row-span](https://user-images.githubusercontent.com/16436270/81493829-5cb60e80-92c1-11ea-8e1a-bc8d7afc1eb0.PNG)

And here's the link to Codesandbox 👉 https://codesandbox.io/s/flattening-row-with-react-table-prkqb?file=/src/App.tsx

Thank you @liberza 🙌

**Reply by rachet-khanal** [reply]
Thanks for the solution. Also looking to find if there is a way we can limit the rowSpan by value like ID of the data?

**Reply by aseph8083** [reply]
i have problem in "isRowSpanned" show message "Property 'isRowSpanned' does not exist on type 'Cell<Data, any>'"

**Reply by Zinoberous** [reply]
@aseph8083 cast everything as any for TypeScript!

@deadcoder0904 nice soluten but there is a problem with grouping, also you can do it a little bit simplier without rowSpanHeaders:

```
<tbody {...getTableBodyProps()}>
  {
    rows.map((row: any, i) => {
      prepareRow(row);

      const topRow: any = rows[i-1];

      for (let j = 0; j < row.allCells.length; j++) {
        let cell = row.allCells[j];

        if (!cell.column.enableRowSpan ||isNull(topRow) || topRow.isGrouped || row.isGrouped) {
          cell.rowSpan = 1;
          cell.isRowSpanned = false;
          continue;
        }

        const getMergeTopCell = (ri: number, ci: number): any => {
          const cell: any = rows[ri].allCells[ci];

          const topRow: any = rows[ri-1];
          const topCell: any = topRow?.allCells[ci];

          if (isNull(topRow) || topRow.isGrouped || JSON.stringify(topCell.value) !== JSON.stringify(cell.value)) {
            return cell
          } else {
            return getMergeTopCell(ri-1, ci);
          }
        }

        let topCell = topRow.allCells[j];

        if (JSON.stringify(topCell.value) === JSON.stringify(cell.value)) {
          getMergeTopCell(i, j).rowSpan += 1;
          cell.isRowSpanned = true;
        } else {
          cell.rowSpan = 1;
          cell.isRowSpanned = false;
        }
      }

      return null;
    })
  }
  {
    rows.map((row: any) => {
      return (
        <tr {...row.getRowProps()}>
          {
            row.cells.map((cell: any) => {
              if (cell.isRowSpanned) return null;
              else
                return (
                  <td
                    rowSpan={cell.rowSpan}
                    {...cell.getCellProps()}
                    title={!isNull(cell.column.Tooltip) ? cell.column.Tooltip(cell) : undefined}
                    style={!isNull(cell.column.Style) ? cell.column.Style(cell) : undefined}
                  >
                    {
                      cell.isGrouped
                      ? (
                          <>
                            <span {...row.getToggleRowExpandedProps()}>
                              {row.isExpanded ? <AiFillCaretDown /> : <AiFillCaretRight />}
                            </span>{' '}
                            {cell.render('Cell')}
                            {' '}
                            ({row.subRows.length})
                          </>
                        )
                      : cell.isAggregated
                        ? cell.render('Aggregated')
                        : cell.isPlaceholder
                          ? null 
                          : cell.render('Cell')
                    }
                  </td>
                );
            })
          }
        </tr>
      );
    })
  }
</tbody>
```

**Reply by WishYouANiceDay** [reply]
Thanks!

**Reply by manoharanamphora** [reply]
how to do the same in v8

**Reply by df-peejay** [reply]
Any road map to support the same rowspan in v8?

**Reply by MDziurakh** [reply]
@tannerlinsley, hi! We really need some documentation about rowSpan! Please, help!

**Reply by DatTN95** [reply]
can you show me a example in v8?

**Reply by syldouss** [reply]
@DatTN95 
Based previous propositions from @deadcoder0904 and @Zinoberous, here is an implementation in V8

Here is the Table component

```javascript
// Table.tsx
import {
  ColumnDef,
  Row,
  flexRender,
  getCoreRowModel,
  useReactTable,
} from "@tanstack/react-table";

const renderRows = (rows: Row<any>[]) => {
  console.log("rows start", rows);
  rows.map((row: any, i: number, rows) => {
    const topRow: any = rows[i - 1];

    for (let j = 0; j < row.getVisibleCells().length; j++) {
      let cell = row.getVisibleCells()[j];

      if (
        !cell.column.columnDef.meta?.enableRowSpan ||
        !topRow ||
        topRow?.getIsGrouped() ||
        row?.getIsGrouped()
      ) {
        cell.rowSpan = 1;
        cell.isRowSpanned = false;
        continue;
      }

      const getMergeTopCell = (ri: number, ci: number): any => {
        const cell: any = (rows[ri] as any).getVisibleCells()[ci];

        const topRow: any = rows[ri - 1];
        const topCell: any = topRow?.getVisibleCells()[ci];

        if (
          !topRow ||
          topRow.getIsGrouped() ||
          JSON.stringify(topCell.getValue()) !== JSON.stringify(cell.getValue())
        ) {
          return cell;
        } else {
          return getMergeTopCell(ri - 1, ci);
        }
      };

      let topCell = topRow.getVisibleCells()[j];

      if (
        JSON.stringify(topCell.getValue()) === JSON.stringify(cell.getValue())
      ) {
        getMergeTopCell(i, j).rowSpan += 1;
        cell.isRowSpanned = true;
      } else {
        cell.rowSpan = 1;
        cell.isRowSpanned = false;
      }
    }

    return null;
  });

  console.log("rows prepared", rows);

  return rows.map((row: any) => {
    return (
      <tr key={row.id}>
        {row.getVisibleCells().map((cell: any) => {
          if (cell.isRowSpanned) return null;
          else
            return (
              <td
                key={cell.id}
                rowSpan={cell.rowSpan}
                style={{ border: "1px solid black", padding: "4px" }}
              >
                {flexRender(cell.column.columnDef.cell, cell.getContext())}
              </td>
            );
        })}
      </tr>
    );
  });
};

type TableProps = {
  columns: ColumnDef<any, any>[];
  data: object[];
};

export default function Table({ columns, data }: TableProps) {
  const table = useReactTable({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
  });

  return (
    <table
      style={{
        borderCollapse: "collapse",
      }}
    >
      <thead className="">
        {table.getHeaderGroups().map((headerGroup: any) => (
          <tr key={headerGroup.id} style={{}}>
            {headerGroup.headers.map((header: any) => (
              <th
                key={header.id}
                scope="col"
                style={{ border: "1px solid black", padding: "4px" }}
              >
                {header.isPlaceholder
                  ? null
                  : flexRender(
                      header.column.columnDef.header,
                      header.getContext()
                    )}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody className="">{renderRows(table.getRowModel().rows)}</tbody>
    </table>
  );
}
```

And here is how to call it with props:

```javascript
// App.tsx
import "./styles.css";
import Table from "./Table";
import { createColumnHelper } from "@tanstack/react-table";

export type Player = {
  team: string;
  firstName: string;
  lastName: string;
};

const data = [
  {
    team: "LA Lakers",
    firstName: "Lebron",
    lastName: "James",
  },
  { team: "LA Lakers", firstName: "Anthony", lastName: "Davis" },
  { team: "Boston Celtics", firstName: "Jayson", lastName: "Tatum" },
  { team: "Boston Celtics", firstName: "Jaylen", lastName: "Brown" },
  { team: "NY Knicks", firstName: "Jalen", lastName: "Brunson" },
];

const columnHelper = createColumnHelper<Player>();
const columns = [
  columnHelper.accessor("team", {
    header: () => "Team",
    meta: { enableRowSpan: true },
  }),
  columnHelper.accessor("firstName", {
    header: () => "Firstname",
  }),
  columnHelper.accessor("lastName", {
    header: () => "Lastname",
  }),
];

export default function App() {
  return (
    <div className="App">
      <h1>Here are the ball teams players</h1>
      <Table columns={columns} data={data} />
    </div>
  );
}
```

You can find sandbox here : https://codesandbox.io/p/sandbox/tanstack-react-table-with-rowspan-chmh76

**Reply by DatTN95** [reply]
tks bro

**Reply by michaelsawyers** [reply]
This only works when the data rows that are intended to be grouped together are adjacent. So either the original data has the rows that should span together as siblings, or the table is sorted 'asc' or 'desc' on the column to group cells.

Nonetheless, thanks for sharing your V8 solution!

**Reply by kapoor17** [reply]
You are a life saver. Was trying to implement this for a few days now. Thanks a lot!!

**Reply by mindkeeper** [reply]
Thanks a lot dude, you are really a life saver!

**Reply by AdnanTheExcellent** [reply]
Thanks!. Does anyone have updated code without using `any`? es-lint is complaining about the `meta: { enableRowSpan: true }` line not being declared in `ColumnMeta` and all the other `any` types.

**Reply by michaelsawyers** [reply]
@AdnanTheExcellent Take a look at these docs about declaration merging for custom features https://tanstack.com/table/v8/docs/guide/custom-features#adding-a-custom-feature

**Comment by deadcoder0904** [comment]
@liberza I found 1 problem with your solution. Guess it might be an edge-case. But here's the problem.

I have 2 columns that I have to `rowSpan`. The 1st column & the 2nd one. The 1st column is unique in names but the 2nd one isn't.

As the 2nd one is a `ratings` column & `ratings` are sometimes same while the 1st one remains unique.

Here's a visual:

| Name | Ratings | Animals | Places |
|------|---------|---------|--------|
| A    | 5       |         |        |
| A    | 5       |         |        |
| B    | 4.9     |         |        |
| B    | 4.9     |         |        |
| C    | 4.9     |         |        |
| C    | 4.9     |         |        |

So the `Name` column spans perfectly with 2 rows spanning each.

The problem comes while spanning `Ratings`. The 1st two rows span 2. But the 2nd two rows span 4 as `4.9` is same for 4 rows as they are besides. I'm trying to keep track of the 1st unique column ,i.e, `Name` & span the `Ratings` column accordingly. But haven't successfully done it yet.

Let me know if you have any solutions in this regard? Also, keep this use-case in mind while making the plugin, I guess it'll be a common one. Just a hunch :)

**Reply by nleve** [reply]
@deadcoder0904 I finally got around to creating an updated example with how I am doing it now. See here: https://codesandbox.io/s/vigorous-wescoff-nup6e

In order to get the functionality you are looking for, just uncomment the code that refers to `parentBoundary`.

**Reply by nleve** [reply]
Also now that I've got the implementation in a better state, see https://github.com/tannerlinsley/react-table/pull/2534 for progress on adding this as a base plugin.

**Reply by tvone** [reply]
> @liberza I found 1 problem with your solution. Guess it might be an edge-case. But here's the problem.
> 
> I have 2 columns that I have to `rowSpan`. The 1st column & the 2nd one. The 1st column is unique in names but the 2nd one isn't.
> 
> As the 2nd one is a `ratings` column & `ratings` are sometimes same while the 1st one remains unique.
> 
> Here's a visual:
> 
> Name	Ratings	Animals	Places
> A	5		
> A	5		
> B	4.9		
> B	4.9		
> C	4.9		
> C	4.9		
> So the `Name` column spans perfectly with 2 rows spanning each.
> 
> The problem comes while spanning `Ratings`. The 1st two rows span 2. But the 2nd two rows span 4 as `4.9` is same for 4 rows as they are besides. I'm trying to keep track of the 1st unique column ,i.e, `Name` & span the `Ratings` column accordingly. But haven't successfully done it yet.
> 
> Let me know if you have any solutions in this regard? Also, keep this use-case in mind while making the plugin, I guess it'll be a common one. Just a hunch :)

Do you have any solution for v8?

**Comment by Kuncheria-Kuruvilla** [comment]
Is there any way by which we can have row span on the table header, like the one below 

![image](https://user-images.githubusercontent.com/7982559/95952274-6ebe4f00-0e15-11eb-8ec2-0ab9032df3f9.png)

**Reply by timbakkum** [reply]
wondering about this as well, any ideas on how to achieve this?

**Reply by ljubomirsinadinovski** [reply]
Same here... how did u manage to achieve this, can someone please give some reference to solution or code example? @timbakkum or @Kuncheria-Kuruvilla

**Reply by ljubomirsinadinovski** [reply]
I managed to make a hack like this, but i don't know if this is going to have limitations in the future:

Inside of the table structure i modified the returning <th> into:
``` 
                  <th
                    {...column.getHeaderProps()}
                    rowSpan={`${column.rowSpan ?? 1}`}
                    style={column.displayNone ? { display: 'none' } : {}}
                  >
                    {column.render('Header')}
                  </th>
```

While having in the data structure (JSON object containing the Header data) something like this:
```
{
    Header: 'Header Name That has spanRow=3',
    columns: [
      {
        Header: 'Not visible header 2',
        displayNone: true,
        columns: [{ Header: 'Not visible header 3', accessor: 'theAccessor', displayNone: true }],
      },
    ],
    rowSpan: '3',
},
```

**Reply by aliabdollahy** [reply]
thanks! it worked for me.

**Reply by noorqidam** [reply]
> I managed to make a hack like this, but i don't know if this is going to have limitations in the feature:
> 
> Inside of the table structure i modified the returning into:
> 
> ```
>                   <th
>                     {...column.getHeaderProps()}
>                     rowSpan={`${column.rowSpan ?? 1}`}
>                     style={column.displayNone ? { display: 'none' } : {}}
>                   >
>                     {column.render('Header')}
>                   </th>
> ```
> 
> While having in the data structure (JSON object containing the Header data) something like this:
> 
> ```
> {
>     Header: 'Header Name That has spanRow=3',
>     columns: [
>       {
>         Header: 'Not visible header 2',
>         displayNone: true,
>         columns: [{ Header: 'Not visible header 3', accessor: 'theAccessor', displayNone: true }],
>       },
>     ],
>     rowSpan: '3',
> },
> ```

can you provide an example in Codesandbox because I'm having trouble using rowspan
![image](https://user-images.githubusercontent.com/45481403/184595205-275f0eab-0576-4743-90cb-98e8c2a318ae.png)

**Reply by ljubomirsinadinovski** [reply]
did u manage to solve it? Please make the code sandbox and I will write here what should you change. It saves much more time for me. Thanks

**Reply by ghost** [reply]
@tannerlinsley please help me!

**Comment by dan1994** [comment]
I was wondering if there's a way to use `rowSpan` along with `useResizeableColumns`. According to the docs, `useResizeableColumns` works only when using non-table elements. Is there a way to emulate behavior like `rowSpan` in that state?

**Comment by bad4iz** [comment]
![image](https://user-images.githubusercontent.com/16918043/116438189-451fbf00-a85f-11eb-833d-315ff97d8056.png)
how to do this header table?

**Comment by tushar1998** [comment]
@deadcoder0904  Is there a way to achieve this for colspan ? 
<img width="605" alt="Screenshot 2023-08-30 at 11 49 42 AM" src="https://github.com/TanStack/table/assets/22202489/9f78e626-9685-480d-88d2-e104764d1816">

**Reply by deadcoder0904** [reply]
ask chatgpt. haven't used this in a while but chatgpt should be able to answer :)

**Reply by tushar1998** [reply]
Actually I solved this issue by taking reference of parent columns length and checking for undefined values in table. Not a straightforward thing but achievedable

**Reply by itskazemk** [reply]
hey bro, can you share your code?

**Reply by tushar1998** [reply]
I hope this gets you an idea for the solution
```
if (value !== undefined) {
	return (
		<td
		  key={cell.id}
		  colSpan={row.index === 0 ? cell.column.parent?.columns.length : 1}
	        >
		  {children}
		</td>
         );
}
```

**Comment by df-peejay** [comment]
What about the support for the same in v8? Is there any roadmap for this?

**Comment by Roman86** [comment]
Hey everyone! Just solved the header rowspan problem and packaged it into a tiny npm library. It's really simple to use — you can rowspan both the header and footer, with table and non-table tr/th elements. Check it out!

```
npm i tanstack-table-header-rowspan
```

https://www.npmjs.com/package/tanstack-table-header-rowspan

it's about header and footer, not data rows. I probably could implement similar approach with data rows, but need to think more about handling pagination, virtualization and dynamic rows height
